qgsquick/from_qgis/quickgui/attributes/qgsquickattributecontroller.cpp

```diff
@@ -0,0 +1,879 @@
+        int fieldIndex = editorField->idx();
+        if ( fieldIndex < 0 || fieldIndex >= layer->fields().size() )
+        {
+          qDebug() << "Invalid fieldIndex for editorField!";
+          continue;
+        }
+        QgsField field = layer->fields().at( fieldIndex );
+        QStringList expressions;
+        QStringList descriptions;
+        QString expression = field.constraints().constraintExpression();
+
+        if ( !expression.isEmpty() )
+        {
+          descriptions << field.constraints().constraintDescription();
+          expressions << field.constraints().constraintExpression();
+        }
+
+        if ( field.constraints().constraints() & QgsFieldConstraints::ConstraintNotNull )
+        {
+          descriptions << QObject::tr( "Not NULL" );
+        }
+
+        if ( field.constraints().constraints() & QgsFieldConstraints::ConstraintUnique )
+        {
+          descriptions << QObject::tr( "Unique" );
+        }
+
+        const QString groupName = container->isGroupBox() ? container->name() : QString();
+
+        // mConstraints.insert( item, field.constraints() );
+        std::shared_ptr<QgsQuickFormItemData> formItemData =
+          std::shared_ptr<QgsQuickFormItemData>(
+            new QgsQuickFormItemData(
+              fieldUuid,
+              field,
+              groupName,
+              parentTabRow,
+              QgsQuickFormItem::Field,
+              layer->attributeDisplayName( fieldIndex ),
+              !layer->editFormConfig().readOnly( fieldIndex ),
+              layer->editorWidgetSetup( fieldIndex ),
+              fieldIndex,
+              field.constraints(),
+              parentVisibilityExpressions // field doesn't have visibility expression itself
+            )
+          );
+
+        mFormItemsData[formItemData->id()] = formItemData;
+
+
+        items.append( fieldUuid );
+        break;
+      }
+
+      case QgsAttributeEditorElement::AeTypeRelation:
+        // todo
+        break;
+
+      case QgsAttributeEditorElement::AeTypeInvalid:
+        // todo
+        break;
+
+      case QgsAttributeEditorElement::AeTypeQmlElement:
+        // todo
+        break;
+
+      case QgsAttributeEditorElement::AeTypeHtmlElement:
+        // todo
+        break;
+    }
+  }
+}
+
+void QgsQuickAttributeController::createTab( QgsAttributeEditorContainer *container )
+{
+  Q_ASSERT( container );
+
+  int tabRow = mTabItems.size();
+
+  QgsExpression expr;
+  if ( container->visibilityExpression().enabled() )
+  {
+    expr = container->visibilityExpression().data();
+  }
+
+  QVector<QUuid> formItemsUuids;
+  flatten( container, tabRow, QString(), formItemsUuids );
+
+  std::shared_ptr<QgsQuickTabItem> tabItem( new QgsQuickTabItem( tabRow,
+      container->name(),
+      formItemsUuids,
+      expr )
+                                          );
+
+  mTabItems.push_back( tabItem );
+}
+
+void QgsQuickAttributeController::clearAll()
```

Comment:
how is this different from forceClean() ? is it possible to merge them to a single function?

Reply:
this is private function --- helper to clean the internal members. forceClean is public way how to reset